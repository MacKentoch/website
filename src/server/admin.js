import express from 'express'
import objection from 'objection'
import ObjectionRest from 'objection-rest'
import bodyParser from 'body-parser'
import Trainer from 'server/models/Trainer'
import Training from 'server/models/Training'
import Path from 'server/models/Path'
import Course from 'server/models/Course'
import * as database from 'server/services/database'
import path from 'path'

const app = express()

app.use(bodyParser.json())
database.connect()

ObjectionRest(objection)
  .routePrefix('/api')
  .pluralizer(x => x)
  .addModel(Trainer)
  .addModel(Path)
  .addModel(Training)
  .addModel(Course)
  .generate(app)

app.use(express.static(path.join(__dirname, '../../public-admin/dist')))
app.listen(3000)
