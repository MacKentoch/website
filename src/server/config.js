import path from 'path'
import convict from 'convict'

const config = convict({
  env: {
    doc: 'The application environment',
    format: [
      'production',
      'development',
      'test',
      'browser.development',
    ],
    default: 'development',
    env: 'NODE_ENV',
  },
  server: {
    port: {
      doc: 'The server port number',
      format: 'port',
      default: 8000,
      env: 'PORT',
    },
    logFormat: {
      doc: 'The morgan log format to use',
      format: [
        'dev',
        'combined',
        'common',
        'short',
        'tiny',
        '',
      ],
      default: 'dev',
    },
  },
})

const env = config.get('env')
config.loadFile(path.join(__dirname, `../../config/${env}.json`))
config.validate()

export default config
