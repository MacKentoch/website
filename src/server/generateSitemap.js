import sm from 'sitemap'
import Training from 'server/models/Training'
import TrainingSession from 'server/models/TrainingSession'
import { trainingRoute } from 'modules/routePaths'

async function generateSitemap() {
  const trainings = await Training.query().orderByRaw(
    'updated_at desc, id desc',
  )
  const sessions = await TrainingSession.query()
    .eager('[training, location]')
    .whereRaw("start_date > now() + interval '14 day'")
    .orderBy('start_date', 'desc')
  const sessionLinks = await Promise.all(
    sessions.map(session => session.link()),
  )

  const sitemap = sm.createSitemap({
    hostname: 'https://www.smooth-code.com',
    cacheTime: 600000,
    urls: [
      { url: '/', changefreq: 'weekly', priority: 1 },
      { url: '/trainings', changefreq: 'weekly', priority: 0.8 },
      { url: '/story', priority: 0.6 },
      { url: '/contact', priority: 0.6 },
      ...trainings.map(training => ({
        url: trainingRoute(training.slug),
        lastmodISO: training.updated_at.toISOString(),
        changefreq: 'weekly',
        priority: 0.9,
      })),
      ...sessions.map((session, index) => ({
        url: sessionLinks[index],
        lastmodISO: session.updated_at.toISOString(),
        changefreq: 'weekly',
        priority: 0.7,
      })),
    ],
  })

  return new Promise((resolve, reject) => {
    sitemap.toXML((err, xml) => {
      if (err) {
        reject(err)
        return
      }

      resolve(xml.toString())
    })
  })
}

export default generateSitemap
