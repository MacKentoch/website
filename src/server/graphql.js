import Path from 'server/models/Path'
import Training from 'server/models/Training'
import Trainer from 'server/models/Trainer'
import TrainingSession from 'server/models/TrainingSession'
import { GraphQLDate } from 'graphql-iso-date'
import { makeExecutableSchema } from 'graphql-tools'
import graphqlFields from 'graphql-fields'

const resolvers = {
  Date: GraphQLDate,
}

export const schema = makeExecutableSchema({
  typeDefs: `
    scalar Date

    type Path {
      id: ID!
      title: String
      color: String
      icon: String
      trainings: [Training]
    }

    type Course {
      id: ID!
      title: String
      outline: String
      path: Path
    }

    type Location {
      id: ID!
      name: String
      address: String
      city: String
      zipcode: String
      country: String
    }

    type Training {
      slug: ID!
      title: String
      abstract: String
      icon: String
      link: String
      duration: Int
      coursePrice: Int
      dayPrice: Int
      intraPrice: Int
      interPrice: Int
      path: Path
      description: String
      objectives: String
      prerequisites: String
      courses: [Course]
      trainers: [Trainer]
      sessions: [Session]
    }

    type Trainer {
      slug: ID!
      fullName: String
      description: String
      picture: String
      link: String
      trainings: [Training]
    }

    type Session {
      id: ID!
      created_at: Date
      start_date: Date
      end_date: Date
      location: Location
      link: String
      training: Training
    }

    type Query {
      paths: [Path]
      trainings: [Training]
      training(slug: ID!): Training
      trainingSession(id: ID!): Session
      trainer(slug: ID!): Trainer
    }
  `,
  resolvers,
})

const normalizeQuery = pathOrQuery =>
  typeof pathOrQuery === 'string'
    ? {
        path: pathOrQuery,
        modifiers: {},
      }
    : pathOrQuery

class Eager {
  paths = []
  modifiers = {}

  add(pathOrQuery) {
    const query = normalizeQuery(pathOrQuery)
    this.paths = [...this.paths, query.path]
    this.modifiers = { ...this.modifiers, ...query.modifiers }
  }

  toQuery() {
    if (this.paths.length === 0) return null
    return {
      path: `[${this.paths.join(',')}]`,
      modifiers: this.modifiers,
    }
  }
}

const joinQuery = (source, join) => {
  const sourceQuery = normalizeQuery(source)
  const joinQuery = normalizeQuery(join)

  return {
    path: `${sourceQuery.path}${joinQuery ? `.${joinQuery.path}` : ''}`,
    modifiers: {
      ...sourceQuery.modifiers,
      ...(joinQuery ? joinQuery.modifiers : {}),
    },
  }
}

const eagerResolver = {
  paths(fields) {
    const eager = new Eager()
    if (fields.trainings)
      eager.add(
        joinQuery(
          {
            path: 'trainings',
            modifiers: {
              orderByRank(builder) {
                return builder.orderBy('trainings.rank', 'asc')
              },
            },
          },
          eagerResolver.trainings(fields.trainings),
        ),
      )

    return eager.toQuery()
  },
  trainings(fields) {
    const eager = new Eager()
    if (fields.path) eager.add('path')
    if (fields.trainers) eager.add('trainers')
    if (
      fields.courses ||
      fields.duration ||
      fields.intraPrice ||
      fields.interPrice
    )
      eager.add(
        joinQuery('courses', eagerResolver.courses(fields.courses || {})),
      )
    if (fields.sessions)
      eager.add(
        joinQuery(
          {
            path: 'sessions(liveSessions)',
            modifiers: {
              liveSessions(builder) {
                return builder
                  .whereRaw(
                    "training_sessions.start_date > now() + interval '1 day'",
                  )
                  .orderBy('training_sessions.start_date', 'asc')
                  .limit(3)
              },
            },
          },
          eagerResolver.sessions(fields.sessions),
        ),
      )
    return eager.toQuery()
  },
  courses(fields) {
    const eager = new Eager()
    if (fields.path) eager.add('path')
    return eager.toQuery()
  },
  sessions(fields) {
    const eager = new Eager()
    if (fields.location || fields.link) eager.add('location')
    if (fields.training || fields.link) {
      eager.add(
        joinQuery('training', eagerResolver.trainings(fields.training || {})),
      )
    }
    return eager.toQuery()
  },
  trainers(fields) {
    const eager = new Eager()
    if (fields.trainings)
      eager.add(
        joinQuery('trainings', eagerResolver.trainings(fields.trainings)),
      )
    return eager.toQuery()
  },
}

const enhanceQuery = (query, eagerQuery) => {
  if (eagerQuery) return query.eager(eagerQuery.path, eagerQuery.modifiers)
  return query
}

export const rootValue = {
  async paths(args, obj, context) {
    return enhanceQuery(
      Path.query().orderBy('paths.rank', 'asc'),
      eagerResolver.paths(graphqlFields(context)),
    )
  },
  async training({ slug }, obj, context) {
    return enhanceQuery(
      Training.query().where({ 'trainings.slug': slug }).first(),
      eagerResolver.trainings(graphqlFields(context)),
    )
  },
  async trainingSession({ id }, obj, context) {
    return enhanceQuery(
      TrainingSession.query().where({ 'training_sessions.id': id }).first(),
      eagerResolver.sessions(graphqlFields(context)),
    )
  },
  async trainer({ slug }, obj, context) {
    return enhanceQuery(
      Trainer.query().where({ 'trainers.slug': slug }).first(),
      eagerResolver.trainers(graphqlFields(context)),
    )
  },
}
