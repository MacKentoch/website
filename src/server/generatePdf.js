import tmp from 'tmp-promise'
import { createReadStream } from 'fs'
import { exec } from 'mz/child_process'
import config from 'server/config'

async function generatePdf(url) {
  const { path: tmpFile } = await tmp.file()
  const chromeBin = config.get('chrome.bin')
  await exec(
    `"${chromeBin}" --headless --disable-gpu --no-sandbox --print-to-pdf=${tmpFile} ${url}`,
  )
  return createReadStream(tmpFile)
}

export default generatePdf
