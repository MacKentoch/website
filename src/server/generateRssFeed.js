import RSS from 'rss'
import { gql } from 'server/graphql'
import { articlesRoute } from 'modules/routePaths'
import { completeUrl } from 'modules/urlUtil'
import { absClUrl } from 'modules/cloudinary'

async function generateRss() {
  const { data, errors } = await gql`
    {
      articles {
        title
        link
        published_at
        author {
          name
        }
        tags {
          name
        }
        feature_image {
          url
        }
      }
    }
  `

  if (errors) {
    console.error(errors) // eslint-disable-line no-console
    throw new Error('Error during rss generation')
  }

  const feed = new RSS({
    title: 'Smooth Code Actualité',
    description: 'Actualité JavaScript, React et Node.js.',
    generator: 'Smooth Code',
    feed_url: completeUrl('/feed.xml'),
    site_url: completeUrl(articlesRoute()),
    image_url: absClUrl('ujjxt9mthnloobzvxfpa', 'q_auto'),
    managingEditor: 'contact@smooth-code.com',
    language: 'fr',
  })

  data.articles.forEach(article =>
    feed.item({
      title: article.title,
      description: article.custom_excerpt,
      url: completeUrl(article.link),
      categories: article.tags.map(tag => tag.name),
      author: article.author.name,
      date: article.published_at,
    }),
  )

  return feed.xml()
}

export default generateRss
