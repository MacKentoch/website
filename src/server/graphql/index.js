import { buildSchema } from 'graphql'

export const schema = buildSchema(`
  type Training {
    name: String
  }

  type Query {
    training(id: ID!): Training
  }
`)

export const rootValue = {
  training() {
    return { name: 'foo' }
  },
}
