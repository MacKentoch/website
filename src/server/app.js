import path from 'path'
import Koa from 'koa'
import morgan from 'koa-morgan'
import favicon from 'koa-favicon'
import serve from 'koa-static'
import conditional from 'koa-conditional-get'
import etag from 'koa-etag'
import compress from 'koa-compress'
import bodyParser from 'koa-bodyparser'
import error from 'koa-error'
import Router from 'koa-router'
import mount from 'koa-mount'
import Training from 'server/models/Training'
import { graphqlKoa, graphiqlKoa } from 'graphql-server-koa'
import config from 'server/config'
import ssr from 'server/middlewares/ssr'
import redirect from 'server/middlewares/redirect'
import generateSitemap from 'server/generateSitemap'
import generatePdf from 'server/generatePdf'
import sendEmail from 'server/email/sendEmail'
import { schema, rootValue } from 'server/graphql'
import { trainingPdfRoute, trainingPrintRoute } from 'modules/routePaths'

const app = new Koa()
const router = new Router()

app.use(
  redirect([
    {
      match: /^\/trainings\/formation-javascript-es2017/,
      redirect: '/formations/javascript-es2017',
    },
    {
      match: /^\/trainings\/formation-nodejs/,
      redirect: '/formations/nodejs',
    },
    {
      match: /^\/trainings\/formation-react/,
      redirect: '/formations/react',
    },
    {
      match: /^\/trainings\/formation-rxjs/,
      redirect: '/formations/rxjs',
    },
    {
      match: /^\/trainings\/formation-graphql/,
      redirect: '/formations/graphql',
    },
    {
      match: /^\/trainings\/formation-jest/,
      redirect: '/formations/react',
    },
    {
      match: /^\/trainings\/formation-initiation-react/,
      redirect: '/formations/react',
    },
    {
      match: /^\/trainings$/,
      redirect: '/formations',
    },
    {
      match: /^\/story/,
      redirect: '/notre-histoire',
    },
    {
      match: /^\/trainers\/greg-berge/,
      redirect: '/formateurs/greg-berge',
    },
    {
      match: /^\/trainers\/adrien-joly/,
      redirect: '/formateurs/adrien-joly',
    },
    {
      match: /^\/trainers\/thomas-jeanneau/,
      redirect: '/formateurs/thomas-jeanneau',
    },
    {
      match: /^\/creer-app-mac-avec-script-shell/,
      redirect: '/articles/creer-app-mac-avec-script-shell',
    },
    {
      match: /^\/developpez-plus-vite-avec-prettier/,
      redirect: '/articles/developpez-plus-vite-avec-prettier',
    },
    {
      match: /^\/pourquoi-react-est-il-si-populaire/,
      redirect: '/articles/pourquoi-react-est-il-si-populaire',
    },
  ]),
)

router.get('/sitemap.xml', async ctx => {
  ctx.response.type = 'xml'
  ctx.response.body = await generateSitemap()
})

router.post('/api/contact', async ctx => {
  const {
    name,
    company,
    email,
    phone,
    message,
    subject = `${ctx.request.body.name} nous a contactés sur smooth-code.com`,
  } = ctx.request.body
  await sendEmail({
    from: email,
    to: 'contact@smooth-code.com',
    subject,
    textContent: `
${subject}

-----

Nom: ${name}

Société: ${company}

Email: ${email}

Téléphone: ${phone}

-----

${message}
`,
  })
  ctx.body = { error: false }
})

router.get(trainingPdfRoute(':slug'), async ctx => {
  const training = await Training.query()
    .where({ slug: ctx.params.slug })
    .first()

  if (!training) {
    const error = new Error('Training not found')
    error.status = 404
    throw error
  }

  ctx.type = 'pdf'
  ctx.attachment(`${training.slug}-smooth-code.pdf`)
  ctx.body = await generatePdf(
    `${config.get('server.externalUrl')}${trainingPrintRoute(training.slug)}`,
  )
})

const PUBLIC = path.join(__dirname, '../../public')

app.use(error())
app.use(bodyParser())
app.use(serve(PUBLIC, { immutable: true, maxage: 31536000000 }))
app.use(compress({ filter: contentType => /text/i.test(contentType) }))
app.use(conditional())
app.use(etag())

if (config.get('env') !== 'test') {
  app.use(morgan(config.get('server.logFormat')))
}

app.use(favicon(path.join(PUBLIC, 'favicon.ico')))
app.use(router.routes())
app.use(
  mount(
    '/graphql',
    graphqlKoa({
      schema,
      rootValue,
    }),
  ),
)
app.use(mount('/graphiql', graphiqlKoa({ endpointURL: '/graphql' })))
app.use(ssr())

export default app
