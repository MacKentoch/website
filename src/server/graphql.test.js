import { useDatabase, factory } from 'server/test'
import { gql } from 'server/graphql'

describe('graphql', () => {
  useDatabase()

  describe('paths', () => {
    beforeEach(async () => {
      const paths = await factory.createMany('path', 5, [
        { title: 'JavaScript' },
      ])
      await factory.createMany('training', 5, {
        path_id: paths[0].id,
      })
    })

    it('should return path without trainings', async () => {
      const result = await gql`
        {
          paths {
            title
          }
        }
      `

      expect(result.data.paths[0].title).toBe('JavaScript')
    })

    it('should return path with trainings[duration, title]', async () => {
      const result = await gql`
        {
          paths {
            title
            trainings {
              title
              duration
            }
          }
        }
      `

      expect(result.data.paths[0].trainings.length).toBe(5)
    })

    it('should return path with trainings[duration, title, path]', async () => {
      const result = await gql`
        {
          paths {
            title
            trainings {
              title
              duration
              path {
                title
              }
            }
          }
        }
      `

      expect(result.data.paths[0].trainings.length).toBe(5)
      expect(result.data.paths[0].trainings[0].path.title).toBe('JavaScript')
    })
  })
})
