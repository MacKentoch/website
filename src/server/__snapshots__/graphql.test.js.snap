// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphql trainings should return custom trainer 1`] = `
Object {
  "data": Object {
    "trainer": Object {
      "fullName": "Greg Bergé",
    },
  },
}
`;

exports[`graphql trainings should return custom training 1`] = `
Object {
  "data": Object {
    "training": Object {
      "trainers": Array [
        Object {
          "fullName": "Greg Bergé",
        },
        Object {
          "fullName": "Adrien Joly",
        },
      ],
    },
  },
}
`;

exports[`graphql trainings should return trainings 1`] = `
Object {
  "data": Object {
    "trainings": Array [
      Object {
        "abstract": "Découvrez les nouveautés apportées par ES6 / ES2015, ES2016 et ES2017",
        "cloudinary_id": "training-javascript",
        "description": "JavaScript est un langage en perpétuelle évolution, chaque année des nouveautés y sont apportées. La formation **JavaScript ES2017** se focalise sur les innovations apportées au langage par **ES2015 (ES6)**, **ES2016** et **ES2017**.

Nous présenterons la **destructuration**, les **modules**, les **arrow functions**, les **classes** et les **générateurs**. Nous aborderons aussi la gestion de l’asynchrone de manière élégante avec les **promises** et **async / await**. Côté outillage nous parlerons de **Babel**, **Webpack** et **ESLint** pour vous permettre d'intégrer les nouvelles fonctionnalités dans vos projets.

Un projet sera réalisé en fil rouge. Vous pourrez ainsi mettre en pratique les nouveautés évoquées tout au long de la formation.

### Objectifs
- Découvrir les nouveautés apportées par ES2015 (ES6), ES2016 et ES2017
- Être capable d’utiliser ces nouveautés en production dès le lendemain

### Pré-requis
- Notions de base en développement ou connaissance d'un autre langage de programmation.",
        "duration": 4,
        "name": "JavaScript ES2017",
        "sessions": Array [],
      },
      Object {
        "abstract": "Libérez la puissance du JavaScript côté serveur",
        "cloudinary_id": "nodejs_jxent0",
        "description": "En quelques années Node.js a complètement révolutionné le développement des applications web. Offrant des **performances optimales** et une **flexibilité sans égale**, Node.js a su s'imposer comme le choix de prédilection pour concevoir des applications web. **Basé sur le moteur V8 du navigateur Chrome**, il permet de **faire tourner le même code côté client et côté serveur**. Node.js tire également sa force de son **gestionnaire de paquet \\"npm\\"** qui comptabilise plus de **475 000 modules disponibles**.

Utilisé par tous [**les géants du web**](https://nodejs.org/en/foundation/case-studies/), Node.js est aujourd'hui incontournable. En plus de permettre la **création d'applications web réactives grâce à WebSocket**, Node.js permet d'utiliser les **nouveaux outils de développement en ligne de commande** comme Webpack, Babel ou ESLint.

Dans cette formation vous découvrirez comment écrire du code **JavaScript moderne (ES2017) avec [Babel](https://babeljs.io/)**, profiter de l'écosystème grâce à **[npm](https://www.npmjs.com) et ses modules**,  maîtriser **l'asynchrone avec \`async\` / \`await\`**,  créer un **serveur web avec [Express](https://expressjs.com/) ou [Koa](http://koajs.com/)**,  créer des **interfaces réactives avec [Socket.io](https://socket.io/)** et écrire du **code isomorphique (client et serveur)**, **se connecter à une base de données avec [knex](http://knexjs.org/)**, **tester son code avec [Jest](https://facebook.github.io/jest/)**.

Un projet complet sera réalisé en fil rouge durant toute la formation, cela vous permettra de mettre en pratique toutes les nouveautés qui vous seront présentées.

### Objectifs
- Savoir installer et utiliser Node.js et npm
- Comprendre le fonctionnement de Node.js
- Savoir créer une application web avec Node.js
- Apprendre à structurer son code de manière optimale
- Être capable de tester son code Node.js

### Pré-requis
- Des bases en JavaScript
- Notions de fonctionnement basiques du web (client, server, HTTP,  requêtes, etc...)",
        "duration": 4,
        "name": "Node.js",
        "sessions": Array [],
      },
      Object {
        "abstract": "Apprenez à créer des interfaces web modernes, fluides et interactives",
        "cloudinary_id": "training-react",
        "description": "React est un moteur de rendu développé par Facebook se démarquant de la concurrence grâce à son architecture innovante, efficace et performante. React a maintenant une place incontestable dans l'écosystème JavaScript.


Pendant cette formation, vous découvrirez l'univers React et comment l'utiliser de façon optimale dans vos projets. Nous aborderons tous les aspects de React : sa philosophie, les **components**, les **applications isomorphiques**, mais aussi **comment bien architecturer son application**, l'**optimiser** et la **déboguer**. Nous vous accompagnerons dans l'installation d'un environnement optimal pour le développement en React avec les outils : **Atom**, **Babel**, **Prettier** et **ESLint**.

Un projet complet sera réalisé en fil rouge durant toute la formation pour vous permettre de mettre en pratique chacune des nouveautés présentées.

### Objectifs
- Découvrir React et maîtriser ses particularités
- Construire votre première application React en respectant les bonnes pratiques
- Utiliser les outils de développement adaptés
- Installer, configurer et utiliser les bibliothèques nécessaires
- Savoir comment intégrer React dans vos projets dès le lendemain

### Pré-requis
- Niveau JavaScript intermédiaire",
        "duration": 3,
        "name": "React",
        "sessions": Array [],
      },
      Object {
        "abstract": "Maîtrisez la programmation réactive",
        "cloudinary_id": "training-rxjs",
        "description": "La programmation fonctionnelle est maintenant incontournable dans l'écosystème JavaScript. Mise en avant par Lodash, puis par React et Redux, elle ouvre de nouvelles portes dans le développement web. La programmation Réactive peut être considérée comme l'étape suivante à la programmation fonctionnelle. Popularisée par Angular, le type Observable (cœur de la programmation Réactive) est en cours de standardisation et fera bientôt partie du langage.

La programmation Réactive Fonctionnelle permet de modéliser du code asynchrone sous la forme d'une simple collection avec une dimension temporelle. Très puissante, elle peut-être cependant difficile à appréhender.

RxJS est la bibliothèque la plus complète et la plus populaire pour appliquer ce paradigme. Elle est notamment utilisée dans Angular, ce qui fait d'elle une librairie incontournable.

Dans cette formation vous comprendrez la différence d'approche **pull / push**, le pattern **Observable** et **Observer** et l'**API RxJS dans des cas d'utilisation réels**.

Plusieurs exercices concrets tout au long de la formation vous permettront de mettre en pratique les concepts vus lors de ces deux jours.

### Objectifs
- Comprendre le paradigme de programmation réactive
- Maitriser la librairie RxJS (Développement et Test)
- Savoir appliquer la programmation réactive sur des cas réels

### Pré-requis
- Niveau intermédiaire en JavaScript",
        "duration": 2,
        "name": "RxJS",
        "sessions": Array [],
      },
      Object {
        "abstract": "Créez des API modernes et évolutives avec GraphQL",
        "cloudinary_id": "training-graphql",
        "description": "GraphQL est une spécification de requêtes permettant de structurer les réponses à volonté. Elle donne lieu à une nouvelle manière de concevoir des API. En 2012, Facebook se sentait bridé par les limitations des API REST traditionnelles, et a donc conçu GraphQL, une API caractérisée par un seul point d'accès et une architecture sous forme de graphe.

Dans cette formation vous découvrirez **une présentation rapide de GraphQL**. S'ensuivra un comparatif entre **une API GraphQL et une API REST traditionnelle**. Vous explorerez l'ensemble de la spécification, les **Query**, les **Types**, les **Mutations**, les **Fragments** et apprendrez comment l'utiliser dans vos projets de manière concrète.

Un projet complet sera réalisé en fil rouge durant toute la formation, cela vous permettra de mettre en pratique chacune des nouveautés au fil de la formation.

### Objectifs
- Se familiariser avec GraphQL
- Imaginer et modéliser une API GraphQL
- Déployer et utiliser une API GraphQL dans un environnement JavaScript

### Pré-requis
- Notions de fonctionnement basiques du web (client, server, HTTP,  requêtes, etc...)
- Des bases en Node.js
- Des bases en React",
        "duration": 2,
        "name": "GraphQL",
        "sessions": Array [],
      },
      Object {
        "abstract": "Passez à la vitesse supérieure en matière de test JavaScript",
        "cloudinary_id": "training-jest",
        "description": "Jest est le framework de test développé et  utilisé par Facebook. Créé en 2014, il s'est imposé comme référence en tant que framework de tests JavaScript. Jest permet d'optimiser le temps passé à écrire les tests en fournissant un environnement unifié et performant pour une expérience optimale. Il a notamment popularisé le snapshot testing.

Dans cette formation nous verrons comment utiliser Jest pour **tester vos composants React** et **tester votre code côté serveur (Node.js)**. Nous explorerons ensemble le système d'**assertions**, de **mocking** ainsi que le **snapshot testing**. Nous verrons également comment implémenter une **couverture de code simple et efficace avec codecov**.

Nous implémenterons des tests sur des exemples précis pour  vous permettre de faire le parallèle avec vos projets.

### Objectifs
- Maîtriser les meilleures pratiques de test pour des composants React avec Jest
- Savoir utiliser Jest pour tester des modules Node.js

### Pré-requis
- Notions de base en React
- Notions de base en Node.js",
        "duration": 1,
        "name": "Jest",
        "sessions": Array [],
      },
      Object {
        "abstract": "Créez votre première application React en une journée",
        "cloudinary_id": "react-intro_eiiuhv",
        "description": "Vous avez probablement déjà entendu parlé de React, cette technologie popularisée par Facebook permettant de créer des applications réactives et performantes. React a maintenant une place incontestable dans l'écosystème JavaScript. Le maîtriser vous ouvrira les portes du développement front-end, la compétence la plus demandée sur le marché aujourd'hui.

Grâce à cette initiation, vous découvrirez comment créer une application React, l'architecturer, la tester et l'optimiser. Vous serez capable de le mettre en pratique dès demain sur vos projets personnels ou chez des clients.

Tout au long de la journée, vous serez guidé dans la réalisation d'un projet complet en alternant théorie et pratique.

### Objectifs
- Découvrir React
- Construire votre première application React
- Avoir les bases pour aller plus loin

### Pré-requis
- Niveau JavaScript intermédiaire
- Votre ordinateur portable",
        "duration": 1,
        "name": "Initiation à React",
        "sessions": Array [
          Object {
            "location": Object {
              "city": "Paris",
              "name": "Smooth Code",
            },
            "start_date": "2017-12-06",
          },
        ],
      },
    ],
  },
}
`;
